lib/Semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,311 @@
+//===-- lib/semantics/check-select-stmt.cc --------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "check-select-stmt.h"
+#include "flang/Parser/message.h"
+#include "flang/Semantics/tools.h"
+
+namespace Fortran::semantics {
+
+template<typename T> class SelectCaseHelper {
+
+public:
+  SelectCaseHelper(parser::Messages &messages, int kind, TypeCategory category)
+    : kind_{kind}, category_(category), messages_{messages} {}
+  TypeCategory GetCategory() { return category_; }
+
+  void ReportOverlapCaseValError(parser::CharBlock currSource,
+      parser::CharBlock prevSource, const char *val) {
+    // C1149
+    messages_
+        .Say(currSource,
+            "CASE value %s matches a previous CASE statement"_err_en_US, val)
+        .Attach(
+            prevSource, "Previous CASE statement matching %s"_err_en_US, val);
+  }
+  // Before inserting the value, create a range value val:val and check if
+  // it overlaps any existing value
+  void Insert(T val, parser::CharBlock source) {
+    SelectCaseRangeType rangeValue(val, val, source);
+    parser::CharBlock overlappingSource;
+    if (!IsRangeValueOverlapping(rangeValue, overlappingSource)) {
+      selectCaseRangeVals_.push_back(rangeValue);
+    } else {
+      ReportOverlapCaseValError(source, overlappingSource,
+          Fortran::evaluate::GetValueAsString(val).c_str());
+    }
+  }
+  void Insert(std::optional<T> lowerVal, std::optional<T> upperVal,
+      parser::CharBlock source) {
+    SelectCaseRangeType rangeValue(lowerVal, upperVal, source);
+
+    parser::CharBlock overlappingSource;
+    if (!IsRangeValueOverlapping(rangeValue, overlappingSource)) {
+      selectCaseRangeVals_.push_back(rangeValue);
+    } else {
+      ReportOverlapCaseValError(source, overlappingSource,
+          Fortran::evaluate::GetValuePairAsString(lowerVal, upperVal).c_str());
+    }
+  }
+  template<typename U> void Insert(const U &x, parser::CharBlock source) {
+    if constexpr (std::is_same_v<T, bool>) {
```

Comment:
Long chains of `if constexpr` code defeat the goal of generic programming.

I have a short polymorphic implementation of `SELECT CASE` semantic checking that I will push out for review later today that demonstrates what I think this code should look like.